exploit/utils.js

```diff
@@ -107,283 +107,295 @@ exports.toHex = function (d) {
 // throw when truncating non-zero bits
 exports.trunclt32 = function (num, bits) {
-	if(bits > 32) {
-		throw new Error("can't truncate > 32 bits with trunclt32");
-	}
-	if(Array.isArray(num) && this.assertu64(num)) {
-		if(num[1] !== 0) {
-			throw new Error("high " + (64-bits) + " bits must be clear");
-		}
-		num = this.assertu32(num[0]);
-	} else if(typeof(num) === "number") {
-		num = this.assertu32(num);
-	} else {
-		throw new Error("expected [lo,hi] or u32");
-	}
-	// for some reason, a >> 32 == a. that makes literally no sense.
-	if(bits == 32 ? 0 : (num >> bits) > 0) {
-		throw new Error("number is too large for " + bits + " bits");
-	}
-	return num;
+    if(bits > 32) {
+        throw new Error("can't truncate > 32 bits with trunclt32");
+    }
+    if(Array.isArray(num) && this.assertu64(num)) {
+        if(num[1] !== 0) {
+            throw new Error("high " + (64-bits) + " bits must be clear");
+        }
+        num = this.assertu32(num[0]);
+    } else if(typeof(num) === "number") {
+        num = this.assertu32(num);
+    } else {
+        throw new Error("expected [lo,hi] or u32");
+    }
+    // for some reason, a >> 32 == a. that makes literally no sense.
+    if(bits == 32 ? 0 : (num >> bits) > 0) {
+        throw new Error("number is too large for " + bits + " bits");
+    }
+    return num;
 };
 
 // truncate to less than 64 bits, will always return [lo, hi]
 // throw when truncating non-zero bits
 exports.trunclt64 = function (num, bits) {
-	num = this.pad64(num);
-	if(this.bits <= 32) {
-		return [this.trunclt32(num), 0];
-	}
-	// for some reason, a >> 32 == a. that makes literally no sense.
-	if(bits == 64 ? 0 : (num[1] >> (bits-32)) > 0) {
-		throw new Error("number is too large for " + bits + " bits");
-	}
-	return num;
+    num = this.pad64(num);
+    if(this.bits <= 32) {
+        return [this.trunclt32(num), 0];
+    }
+    // for some reason, a >> 32 == a. that makes literally no sense.
+    if(bits == 64 ? 0 : (num[1] >> (bits-32)) > 0) {
+        throw new Error("number is too large for " + bits + " bits");
+    }
+    return num;
 };
 
 exports.pad64 = function (num) {
-	if(Array.isArray(num)) {
-		return this.assertu64(num);
-	} else if(typeof(num) === "number") {
-		return [this.assertu32(num), 0];
-	} else {
-		throw new Error("expected [lo,hi] or number");
-	}
+    if(Array.isArray(num)) {
+        return this.assertu64(num);
+    } else if(typeof(num) === "number") {
+        return [this.assertu32(num), 0];
+    } else {
+        throw new Error("expected [lo,hi] or number");
+    }
 };
 
 console = {
-	log: function (msg) {
-		exports.send('log', msg);
-	}
+    log: function (msg) {
+        exports.send('log', msg);
+    }
 };
 
 var log = console.log;
 
 exports.log = log;
 
+var hasSmhax = false;
+smhax = {
+    get: function () {
+        return hasSmhax;
+    },
+    set: function (state) {
+        hasSmhax = state;
+    }
+};
+
+exports.smhax = smhax;
```

Comment:
This doesn't really need to be a getter/setter. Just `exports.hasSmhax = false;` and keep it simple.

Reply:
Yes, that isn't necessary at all, just pushed it as-is to discuss other ideas for detecting smhax and/or a numeric version field (maybe a version check that changes an exported smhax field).

Reply:
```
exports.hasSmhax = function () {
    return exports.version >= 301;
}
```
would be fine